app/assets/javascripts/basetypes/DataElement.js

```diff
@@ -1,36 +1,69 @@
 const mongoose = require('mongoose');
 const Code = require('./Code.js');
+const changeCase = require('change-case');
 
 const [Schema] = [mongoose.Schema];
 
-function extendSchema(TSchema, definition, options) {
-  return new Schema(Object.assign({}, TSchema.obj, definition), options);
-}
+function DataElementSchema(add, options) {
+  const extended = new Schema({
+    data_element_codes: { type: [Code] },
+    description: { type: String },
+  }, options);
+
+  if (add) {
+    extended.add(add);
+  }
+
+  extended.methods.get = function get(path, type) {
+    let adhoc;
+    const pathSnake = changeCase.snakeCase(path);
+    if (type) {
+      adhoc = Schema.interpretAsType(path, type, this.schema.options) || Schema.interpretAsType(pathSnake, type, this.schema.options);
+    }
+
+    const schema = this.$__path(path) || this.$__path(pathSnake) || this.schema.virtualpath(path) || this.schema.virtualpath(pathSnake);
+
+    const pieces = path.split('.');
+    const piecesSnake = pathSnake.split('.');
+    let obj = this._doc;
+
+    if (schema instanceof mongoose.VirtualType) {
+      return schema.applyGetters(null, this);
+    }
+
+    for (let i = 0, l = pieces.length; i < l; i += 1) {
+      obj = obj === null || obj === undefined
+        ? undefined
+        : (obj[pieces[i]] || obj[piecesSnake[i]]);
+    }
 
-const DataElementSchema = new Schema({
-  data_element_codes: { type: [Code] },
-  description: { type: String },
-});
-
-// Returns the attribute requested on the data element.
-DataElementSchema.methods.get = function get(attribute) {
-  return this[attribute]();
-};
-
-// Returns all of the codes on this data element.
-DataElementSchema.methods.getCode = function getCode() {
-  return this.data_element_codes.map((code) => {
-    const result = {};
-    result.code = code.code;
-    result.system = code.system;
-    return result;
-  });
-};
-
-// Returns all of the codes on this data element.
-DataElementSchema.methods.codes = function getCode() {
-  return this.data_element_codes;
-};
+    if (adhoc) {
+      obj = adhoc.cast(obj);
+    }
+
+    if (schema) {
+      obj = schema.applyGetters(obj, this);
+    }
+
+    return obj;
+  };
+
+  // Returns all of the codes on this data element.
+  extended.methods.getCode = function getCode() {
```

Comment:
This function and the function below it, 'getCodes()' have duplicate comments. Please elaborate on the differences between the two functions.